core/src/main/java/org/apache/calcite/plan/AbstractRelOptPlanner.java

```diff
@@ -44,18 +44,13 @@
  * Abstract base for implementations of the {@link RelOptPlanner} interface.
  */
 public abstract class AbstractRelOptPlanner implements RelOptPlanner {
-  //~ Static fields/initializers ---------------------------------------------
-
-  /** Regular expression for integer. */
-  private static final Pattern INTEGER_PATTERN = Pattern.compile("[0-9]+");
-
   //~ Instance fields --------------------------------------------------------
 
   /**
    * Maps rule description to rule, just to ensure that rules' descriptions
    * are unique.
    */
-  private final Map<String, RelOptRule> mapDescToRule = new HashMap<>();
+  protected final Map<String, RelOptRule> mapDescToRule = new HashMap<>();
```

Comment:
I think we may need to change it to LinkedHashMap to be consistent with the requirement of HepPlanner.

Reply:
It turns out we still need it to be protected to avoid copying.

Reply:
I think that in most cases the number of rules is not very big so I was thinking that copying vs. `mapDescToRule.values()` is not going to have significant performance overhead in the planning phase, thus, I tend to prefer better encapsulation. Having that said, I do not have any concrete measures to support my claims (just instinct that could be wrong) :)

Reply:
I agree the rules is not very big. But in `onNewClass`, every time we add a new logical/physical operator we have to copy it. It will copy N times depends on how many operators you have. Though not a significant overhead, but I think we still need to avoid multiple times copy.

Reply:
I'm fine with any option so I am leaving the final decision up to you :)